File: app/src/main/java/com/touge/learnsunflower/data/PlantRepository.java
package com.touge.learnsunflower.data;

import java.util.List;

/**
 * @Author Touge
 * @Date 2020/4/6 15:28
 * @Description
 */
public class PlantRepository {
    private static final PlantRepository ourInstance = new PlantRepository();

    public static PlantRepository getInstance() {
        return ourInstance;
    }

    private PlantRepository() {
    }

    public List<Plant> getPlants() {
        return PlantContent.ITEMS;
    }

    public Plant getPlant(String plantId) {
        return PlantContent.ITEM_MAP.get(plantId);
    }
}

File: app/src/main/java/com/touge/learnsunflower/data/PlantRepository.kt
package com.touge.learnsunflower.data

/**
 * @Author Touge
 * @Date 2020/4/8 20:26
 * @Description Repository层
 */
class PlantRepository private constructor(private val plantDao: PlantDao) {

    fun getPlant(plantId: String) = plantDao.getPlant(plantId)

    fun getPlants() = plantDao.getPlants()

    fun getPlantsWithGrowZoneNumber(growZoneNumber: Int) =
            plantDao.getPlantWithGrowZoneNumber(growZoneNumber)

    companion object {

        @Volatile
        private var instance: PlantRepository? = null

        @JvmStatic
        fun getInstance(plantDao: PlantDao): PlantRepository {
            return instance ?: synchronized(this) {
                instance ?: PlantRepository(plantDao).also { instance = it }
            }
        }

    }
}